lightning/src/ln/peer_handler.rs

```diff
@@ -1088,15 +1094,75 @@ impl<Descriptor: SocketDescriptor> PeerManager<Descriptor> {
 			}
 		};
 	}
+/// This function can be called every 30 seconds or so, it will:
+/// disconnect the Peer if awaiting_pong is true
+/// prepare to Ping all the Peers that the PeerManager is responsible for and set awaiting_pong to true
+///
+/// After calling this function one would want to call process_events() in order to finish Pinging the Peer
+/// Then one would want to call do_read_event() in order to reset awaiting_pong
+	pub fn timer_tick_occured(&self) {
+		let mut des_set: Vec<Descriptor> = Vec::new();
+		let mut peers_lock = self.peers.lock().unwrap();
+		let peers = peers_lock.borrow_parts();
+		
+		for (Descriptor, mut Peer) in peers.peers.iter_mut() {
+
+			// If the Peer has an outstanding ping add the offending Peers Descriptor to the vector
+			if Peer.awaiting_pong == true{
+
+				let mut descriptor = Descriptor.clone();
+				des_set.push(descriptor);
+
+			}
+			else {
+			// Otherwise we will be Pinging the Peer and taking appropriate action
+			let ping = msgs::Ping {
+ 				ponglen: 0,
+ 				byteslen: 64,
+ 			};
+			Peer.pending_outbound_buffer.push_back((encode_msg!(ping, 18)));
+			let mut descriptor = Descriptor.clone();
+			self.do_attempt_write_data(&mut descriptor, &mut Peer);
+  			Peer.awaiting_pong = true;
+		
+
+			}
+
+		}
+
+		// Disconnect the offending Peers based on the Descriptors aggregated above
+		for (mut Descriptor) in des_set.iter_mut(){
+			// The following code is a duplicating disconnect_event_internal() except it uses the existing lock instead of taking it
+			let no_connection_possible: bool = true;
```

Comment:
We definitely shouldn't be considering this peer "not connectable" here - if they're just offline we don't want to force-close all our channels with them.